File: app/routes/hr_views.py
# ============================================================================
# HR views — org chart, team PTO calendar, performance reviews.
# ============================================================================

from datetime import date, datetime, timezone
from typing import Optional

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.orm import Session, joinedload

from app.database import get_db
from app.models.payroll import Employee
from app.models.pto import PTORequest, PTORequestStatus
from app.models.reviews import PerformanceReview, ReviewStatus

router = APIRouter(prefix="/api/hr", tags=["hr-views"])


# --- Org chart ---------------------------------------------------------------


@router.get("/org-chart")
def org_chart(include_inactive: bool = False, db: Session = Depends(get_db)):
    """The manager tree from Employee.manager_id.

    Cycles (A manages B manages A — possible because manager_id is
    unconstrained) are broken at the second visit and reported, not
    500'd. Employees whose manager is missing/inactive surface as roots.
    """
    q = db.query(Employee)
    if not include_inactive:
        q = q.filter(Employee.is_active.is_(True))
    employees = q.all()
    by_id = {e.id: e for e in employees}

    children: dict[int, list] = {}
    roots: list[Employee] = []
    for e in employees:
        if e.manager_id and e.manager_id in by_id and e.manager_id != e.id:
            children.setdefault(e.manager_id, []).append(e)
        else:
            roots.append(e)

    cycles: list[int] = []
    visited: set[int] = set()

    def _node(e: Employee) -> dict:
        if e.id in visited:
            cycles.append(e.id)
            return {"id": e.id, "name": e.full_name, "cycle": True, "reports": []}
        visited.add(e.id)
        return {
            "id": e.id,
            "name": e.full_name,
            "role": e.role.value if e.role else None,
            "is_active": bool(e.is_active),
            "reports": [
                _node(c)
                for c in sorted(children.get(e.id, []), key=lambda x: x.full_name)
            ],
        }

    tree = [_node(r) for r in sorted(roots, key=lambda x: x.full_name)]
    # Anyone unreached (a pure cycle disconnected from any root) still shows.
    orphans = [_node(e) for e in employees if e.id not in visited]
    return {"tree": tree + orphans, "cycle_employee_ids": sorted(set(cycles))}


# --- Team PTO calendar -------------------------------------------------------


@router.get("/pto-calendar")
def pto_calendar(
    start: date = Query(...),
    end: date = Query(...),
    db: Session = Depends(get_db),
):
    """Approved (and pending, flagged) PTO overlapping the window."""
    if end < start:
        raise HTTPException(status_code=400, detail="end before start")
    rows = (
        db.query(PTORequest)
        .options(joinedload(PTORequest.employee))
        .filter(
            PTORequest.status.in_(
                [PTORequestStatus.APPROVED, PTORequestStatus.PENDING]
            ),
            PTORequest.start_date <= end,
            PTORequest.end_date >= start,
        )
        .order_by(PTORequest.start_date)
        .all()
    )
    return {
        "start": start.isoformat(),
        "end": end.isoformat(),
        "entries": [
            {
                "employee_id": r.employee_id,
                "employee_name": r.employee.full_name if r.employee else None,
                "start_date": r.start_date.isoformat(),
                "end_date": r.end_date.isoformat(),
                "pto_type": r.pto_type.value if r.pto_type else None,
                "hours": float(r.hours or 0),
                "status": r.status.value,
            }
            for r in rows
        ],
    }


# --- Performance reviews -----------------------------------------------------


class ReviewCreate(BaseModel):
    employee_id: int
    reviewer_id: Optional[int] = None
    period_start: date
    period_end: date
    rating: Optional[int] = None
    goals: Optional[str] = None
    feedback: Optional[str] = None


class ReviewUpdate(BaseModel):
    rating: Optional[int] = None
    goals: Optional[str] = None
    feedback: Optional[str] = None


class AcknowledgeRequest(BaseModel):
    employee_comment: Optional[str] = None


def _review_response(r: PerformanceReview) -> dict:
    return {
        "id": r.id,
        "employee_id": r.employee_id,
        "employee_name": r.employee.full_name if r.employee else None,
        "reviewer_id": r.reviewer_id,
        "reviewer_name": r.reviewer.full_name if r.reviewer else None,
        "period_start": r.period_start.isoformat(),
        "period_end": r.period_end.isoformat(),
        "status": r.status.value if r.status else None,
        "rating": r.rating,
        "goals": r.goals,
        "feedback": r.feedback,
        "employee_comment": r.employee_comment,
        "submitted_at": r.submitted_at.isoformat() if r.submitted_at else None,
        "acknowledged_at": (
            r.acknowledged_at.isoformat() if r.acknowledged_at else None
        ),
    }


def _check_rating(rating):
    if rating is not None and not (1 <= rating <= 5):
        raise HTTPException(status_code=400, detail="rating must be 1-5")


@router.get("/reviews")
def list_reviews(employee_id: int = Query(default=None), db: Session = Depends(get_db)):
    q = db.query(PerformanceReview).options(
        joinedload(PerformanceReview.employee),
        joinedload(PerformanceReview.reviewer),
    )
    if employee_id is not None:
        q = q.filter(PerformanceReview.employee_id == employee_id)
    return [
        _review_response(r)
        for r in q.order_by(PerformanceReview.period_end.desc()).all()
    ]


@router.post("/reviews", status_code=201)
def create_review(data: ReviewCreate, db: Session = Depends(get_db)):
    if not db.query(Employee).filter(Employee.id == data.employee_id).first():
        raise HTTPException(status_code=404, detail="Employee not found")
    if data.period_end < data.period_start:
        raise HTTPException(status_code=400, detail="period end before start")
    _check_rating(data.rating)
    r = PerformanceReview(
        employee_id=data.employee_id,
        reviewer_id=data.reviewer_id,
        period_start=data.period_start,
        period_end=data.period_end,
        rating=data.rating,
        goals=data.goals,
        feedback=data.feedback,
    )
    db.add(r)
    db.commit()
    db.refresh(r)
    return _review_response(r)


@router.put("/reviews/{review_id}")
def update_review(review_id: int, data: ReviewUpdate, db: Session = Depends(get_db)):
    r = db.query(PerformanceReview).filter(PerformanceReview.id == review_id).first()
    if not r:
        raise HTTPException(status_code=404, detail="Review not found")
    if r.status != ReviewStatus.DRAFT:
        raise HTTPException(status_code=400, detail="Only drafts are editable")
    _check_rating(data.rating)
    for field in ("rating", "goals", "feedback"):
        value = getattr(data, field)
        if value is not None:
            setattr(r, field, value)
    db.commit()
    return _review_response(r)


@router.post("/reviews/{review_id}/submit")
def submit_review(review_id: int, db: Session = Depends(get_db)):
    r = db.query(PerformanceReview).filter(PerformanceReview.id == review_id).first()
    if not r:
        raise HTTPException(status_code=404, detail="Review not found")
    if r.status != ReviewStatus.DRAFT:
        raise HTTPException(status_code=400, detail="Already submitted")
    r.status = ReviewStatus.SUBMITTED
    r.submitted_at = datetime.now(timezone.utc)
    db.commit()
    return _review_response(r)


@router.post("/reviews/{review_id}/acknowledge")
def acknowledge_review(
    review_id: int, data: AcknowledgeRequest, db: Session = Depends(get_db)
):
    r = db.query(PerformanceReview).filter(PerformanceReview.id == review_id).first()
    if not r:
        raise HTTPException(status_code=404, detail="Review not found")
    if r.status != ReviewStatus.SUBMITTED:
        raise HTTPException(
            status_code=400, detail="Only submitted reviews can be acknowledged"
        )
    r.status = ReviewStatus.ACKNOWLEDGED
    r.acknowledged_at = datetime.now(timezone.utc)
    r.employee_comment = data.employee_comment
    db.commit()
    return _review_response(r)

File: app/static/js/hr_views.js
/**
 * HR Team — org chart, team PTO calendar, performance reviews.
 *
 * Three tabs over the /api/hr views. The org chart renders the manager
 * tree as nested lists (cycles flagged by the backend surface with a
 * warning). The PTO calendar is a windowed list, and reviews follow a
 * draft -> submitted -> acknowledged lifecycle.
 */
const HRViewsPage = {
    _tab: 'org',

    async render() {
        const employees = await API.get('/employees?active_only=false');
        HRViewsPage._employees = employees;
        return `
            <div class="page-header">
                <h2>HR Team</h2>
                <div style="font-size:10px; color:var(--text-muted);">
                    Org chart, team time-off calendar, and performance reviews
                </div>
            </div>
            <div class="toolbar">
                <button class="btn" id="hr-tab-org" onclick="HRViewsPage.showTab('org')">Org Chart</button>
                <button class="btn" id="hr-tab-pto" onclick="HRViewsPage.showTab('pto')">PTO Calendar</button>
                <button class="btn" id="hr-tab-reviews" onclick="HRViewsPage.showTab('reviews')">Reviews</button>
            </div>
            <div id="hr-tab-content"></div>`;
    },

    async showTab(tab) {
        HRViewsPage._tab = tab;
        ['org', 'pto', 'reviews'].forEach(t => {
            const btn = $(`#hr-tab-${t}`);
            if (btn) btn.classList.toggle('btn-primary', t === tab);
        });
        const box = $('#hr-tab-content');
        if (!box) return;
        try {
            if (tab === 'org') box.innerHTML = await HRViewsPage.renderOrgChart();
            else if (tab === 'pto') box.innerHTML = HRViewsPage.renderPtoControls();
            else box.innerHTML = await HRViewsPage.renderReviews();
            if (tab === 'pto') await HRViewsPage.loadPtoCalendar();
        } catch (e) {
            toast(e.message, 'error');
        }
    },

    // --- Org chart ----------------------------------------------------------

    async renderOrgChart() {
        const data = await API.get('/hr/org-chart');
        let warning = '';
        if (data.cycle_employee_ids.length) {
            warning = `<div style="color:var(--danger); font-size:11px; margin-bottom:8px;">
                Reporting cycle detected involving employee IDs:
                ${data.cycle_employee_ids.join(', ')}</div>`;
        }
        if (!data.tree.length) {
            return warning + '<div class="empty-state"><p>No employees.</p></div>';
        }
        return warning + `<div style="margin-top:8px;">${data.tree.map(n => HRViewsPage.orgNode(n)).join('')}</div>`;
    },

    orgNode(node) {
        const badge = node.cycle
            ? ' <span class="badge badge-void">cycle</span>'
            : (node.is_active === false ? ' <span class="badge badge-void">inactive</span>' : '');
        const role = node.role ? ` <span style="color:var(--text-muted); font-size:10px;">${escapeHtml(node.role)}</span>` : '';
        const kids = (node.reports || []).map(c => HRViewsPage.orgNode(c)).join('');
        return `<div style="padding:3px 0 3px 0;">
            <strong>${escapeHtml(node.name)}</strong>${role}${badge}
            ${kids ? `<div style="margin-left:22px; border-left:1px solid var(--gray-200); padding-left:10px;">${kids}</div>` : ''}
        </div>`;
    },

    // --- PTO calendar -------------------------------------------------------

    renderPtoControls() {
        const today = new Date();
        const start = new Date(today.getFullYear(), today.getMonth(), 1);
        const end = new Date(today.getFullYear(), today.getMonth() + 1, 0);
        const iso = d => d.toISOString().slice(0, 10);
        return `
            <div class="form-grid" style="max-width:480px; margin-top:8px;">
                <div class="form-group"><label>From</label>
                    <input type="date" id="hr-pto-start" value="${iso(start)}"></div>
                <div class="form-group"><label>To</label>
                    <input type="date" id="hr-pto-end" value="${iso(end)}"></div>
            </div>
            <button class="btn btn-primary" onclick="HRViewsPage.loadPtoCalendar()">Refresh</button>
            <div id="hr-pto-results" style="margin-top:12px;"></div>`;
    },

    async loadPtoCalendar() {
        const box = $('#hr-pto-results');
        if (!box) return;
        const start = $('#hr-pto-start')?.value;
        const end = $('#hr-pto-end')?.value;
        if (!start || !end) return toast('Pick a date range', 'error');
        try {
            const data = await API.get(`/hr/pto-calendar?start=${start}&end=${end}`);
            if (!data.entries.length) {
                box.innerHTML = '<p style="color:var(--text-muted); font-size:11px;">No time off in this window.</p>';
                return;
            }
            let html = `<div class="table-container"><table>
                <thead><tr><th>Employee</th><th>Type</th><th>From</th><th>To</th>
                <th class="amount">Hours</th><th>Status</th></tr></thead><tbody>`;
            for (const e of data.entries) {
                html += `<tr>
                    <td>${escapeHtml(e.employee_name || '')}</td>
                    <td>${escapeHtml(e.pto_type || '')}</td>
                    <td>${formatDate(e.start_date)}</td>
                    <td>${formatDate(e.end_date)}</td>
                    <td class="amount">${e.hours}</td>
                    <td>${e.status === 'approved'
                        ? '<span class="badge badge-paid">approved</span>'
                        : '<span class="badge badge-sent">pending</span>'}</td>
                </tr>`;
            }
            box.innerHTML = html + '</tbody></table></div>';
        } catch (e) {
            toast(e.message, 'error');
        }
    },

    // --- Performance reviews ------------------------------------------------

    async renderReviews() {
        const reviews = await API.get('/hr/reviews');
        let html = `<div style="margin:8px 0;">
            <button class="btn btn-primary" onclick="HRViewsPage.newReviewModal()">New Review</button>
        </div>`;
        if (!reviews.length) {
            return html + '<div class="empty-state"><p>No performance reviews yet.</p></div>';
        }
        html += `<div class="table-container"><table>
            <thead><tr><th>Employee</th><th>Reviewer</th><th>Period</th>
            <th>Rating</th><th>Status</th><th>Actions</th></tr></thead><tbody>`;
        for (const r of reviews) {
            const status = r.status === 'acknowledged'
                ? '<span class="badge badge-paid">acknowledged</span>'
                : (r.status === 'submitted'
                    ? '<span class="badge badge-sent">submitted</span>'
                    : '<span class="badge badge-void">draft</span>');
            let actions = '';
            if (r.status === 'draft') {
                actions = `<button class="btn" onclick="HRViewsPage.editReviewModal(${r.id})">Edit</button>
                    <button class="btn" onclick="HRViewsPage.submitReview(${r.id})">Submit</button>`;
            } else if (r.status === 'submitted') {
                actions = `<button class="btn" onclick="HRViewsPage.acknowledgeModal(${r.id})">Acknowledge</button>`;
            }
            html += `<tr>
                <td>${escapeHtml(r.employee_name || '')}</td>
                <td>${escapeHtml(r.reviewer_name || '')}</td>
                <td>${formatDate(r.period_start)} &ndash; ${formatDate(r.period_end)}</td>
                <td>${r.rating != null ? `${r.rating}/5` : ''}</td>
                <td>${status}</td>
                <td>${actions}</td>
            </tr>`;
        }
        return html + '</tbody></table></div>';
    },

    _employeeOptions(selected) {
        return (HRViewsPage._employees || []).map(e =>
            `<option value="${e.id}" ${e.id === selected ? 'selected' : ''}>${escapeHtml(e.first_name)} ${escapeHtml(e.last_name)}</option>`
        ).join('');
    },

    newReviewModal() {
        openModal('New Performance Review', `
            <div class="form-group">
                <label>Employee</label>
                <select id="rev-emp">${HRViewsPage._employeeOptions()}</select>
            </div>
            <div class="form-group">
                <label>Reviewer</label>
                <select id="rev-reviewer"><option value="">(none)</option>${HRViewsPage._employeeOptions()}</select>
            </div>
            <div class="form-group">
                <label>Period start</label>
                <input type="date" id="rev-start">
            </div>
            <div class="form-group">
                <label>Period end</label>
                <input type="date" id="rev-end">
            </div>
            <div class="form-group">
                <label>Rating (1-5, optional)</label>
                <input type="number" id="rev-rating" min="1" max="5">
            </div>
            <div class="form-group">
                <label>Goals</label>
                <textarea id="rev-goals"></textarea>
            </div>
            <div class="form-group">
                <label>Feedback</label>
                <textarea id="rev-feedback"></textarea>
            </div>
            <div class="form-actions">
                <button class="btn btn-secondary" onclick="closeModal()">Cancel</button>
                <button class="btn btn-primary" onclick="HRViewsPage.createReview()">Create</button>
            </div>`);
    },

    async createReview() {
        const start = $('#rev-start')?.value;
        const end = $('#rev-end')?.value;
        if (!start || !end) return toast('Period dates are required', 'error');
        const rating = $('#rev-rating').value;
        try {
            await API.post('/hr/reviews', {
                employee_id: parseInt($('#rev-emp').value, 10),
                reviewer_id: $('#rev-reviewer').value ? parseInt($('#rev-reviewer').value, 10) : null,
                period_start: start,
                period_end: end,
                rating: rating ? parseInt(rating, 10) : null,
                goals: $('#rev-goals').value || null,
                feedback: $('#rev-feedback').value || null,
            });
            closeModal();
            toast('Review created');
            HRViewsPage.showTab('reviews');
        } catch (e) {
            toast(e.message, 'error');
        }
    },

    async editReviewModal(reviewId) {
        try {
            const reviews = await API.get('/hr/reviews');
            const r = reviews.find(x => x.id === reviewId);
            if (!r) return toast('Review not found', 'error');
            openModal('Edit Review (draft)', `
                <div class="form-group">
                    <label>Rating (1-5)</label>
                    <input type="number" id="rev-rating" min="1" max="5" value="${r.rating != null ? r.rating : ''}">
                </div>
                <div class="form-group">
                    <label>Goals</label>
                    <textarea id="rev-goals">${escapeHtml(r.goals || '')}</textarea>
                </div>
                <div class="form-group">
                    <label>Feedback</label>
                    <textarea id="rev-feedback">${escapeHtml(r.feedback || '')}</textarea>
                </div>
                <div class="form-actions">
                    <button class="btn btn-secondary" onclick="closeModal()">Cancel</button>
                    <button class="btn btn-primary" onclick="HRViewsPage.updateReview(${reviewId})">Save</button>
                </div>`);
        } catch (e) {
            toast(e.message, 'error');
        }
    },

    async updateReview(reviewId) {
        const rating = $('#rev-rating').value;
        try {
            await API.put(`/hr/reviews/${reviewId}`, {
                rating: rating ? parseInt(rating, 10) : null,
                goals: $('#rev-goals').value || null,
                feedback: $('#rev-feedback').value || null,
            });
            closeModal();
            toast('Review updated');
            HRViewsPage.showTab('reviews');
        } catch (e) {
            toast(e.message, 'error');
        }
    },

    async submitReview(reviewId) {
        try {
            await API.post(`/hr/reviews/${reviewId}/submit`);
            toast('Review submitted');
            HRViewsPage.showTab('reviews');
        } catch (e) {
            toast(e.message, 'error');
        }
    },

    acknowledgeModal(reviewId) {
        openModal('Acknowledge Review', `
            <div class="form-group">
                <label>Employee comment (optional)</label>
                <textarea id="rev-comment"></textarea>
            </div>
            <div class="form-actions">
                <button class="btn btn-secondary" onclick="closeModal()">Cancel</button>
                <button class="btn btn-primary" onclick="HRViewsPage.acknowledge(${reviewId})">Acknowledge</button>
            </div>`);
    },

    async acknowledge(reviewId) {
        try {
            await API.post(`/hr/reviews/${reviewId}/acknowledge`, {
                employee_comment: $('#rev-comment').value || null,
            });
            closeModal();
            toast('Review acknowledged');
            HRViewsPage.showTab('reviews');
        } catch (e) {
            toast(e.message, 'error');
        }
    },
};
